source/common/config/datasource.h

```diff
@@ -25,6 +32,72 @@ std::string read(const envoy::api::v2::core::DataSource& source, bool allow_empt
  */
 absl::optional<std::string> getPath(const envoy::api::v2::core::DataSource& source);
 
+/**
+ * Callback for async data source.
+ */
+using AsyncDataSourceCb = std::function<void(std::string)>;
+
+class LocalAsyncDataProvider {
+public:
+  LocalAsyncDataProvider(Init::Manager& manager, const envoy::api::v2::core::DataSource& source,
+                         bool allow_empty, Api::Api& api, AsyncDataSourceCb&& callback)
+      : init_target_("LocalAsyncDataProvider", [this, &source, allow_empty, &api, callback]() {
+          callback(DataSource::read(source, allow_empty, api));
+          init_target_.ready();
+        }) {
+    manager.add(init_target_);
+  }
+
+  ~LocalAsyncDataProvider() { init_target_.ready(); }
+
+private:
+  Init::TargetImpl init_target_;
+};
+
+using LocalAsyncDataProviderPtr = std::unique_ptr<LocalAsyncDataProvider>;
+
+class RemoteAsyncDataProvider : public Config::DataFetcher::RemoteDataFetcherCallback {
+public:
+  RemoteAsyncDataProvider(Upstream::ClusterManager& cm, Init::Manager& manager,
+                          const envoy::api::v2::core::RemoteDataSource& source, bool allow_empty,
+                          AsyncDataSourceCb&& callback)
+      : allow_empty_(allow_empty), callback_(std::move(callback)),
+        fetcher_(std::make_unique<Config::DataFetcher::RemoteDataFetcher>(cm, source.http_uri(),
+                                                                          source.sha256(), *this)),
+        init_target_("RemoteAsyncDataProvider", [this]() { start(); }) {
+    manager.add(init_target_);
+  }
+
+  ~RemoteAsyncDataProvider() { init_target_.ready(); }
+
+  // Config::DataFetcher::RemoteDataFetcherCallback
+  void onSuccess(absl::string_view data) {
+    callback_(std::string(data));
```

Comment:
If you're constructing string from string_view anyway then probably just take a `const string&`

Reply:
I will change the callback parameter to `const string&`.

Reply:
I meant you can make onSuccess take const string& as well.